core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
You can have multiple map tasks in one JVM right? they'd be executing simultaneously, using this shared state simultaneously. The issue is really Spark though, where that's definitely going to happen. Same for simply calling `getSplits`. Isn't your current PR already doing the same thing though since it also processes the RDDs in parallel?